packages/tailwindcss/src/variants.ts

```diff
@@ -226,8 +226,49 @@ export class Variants {
     }
 
     let compareFn = this.compareFns.get(aOrder)
-    if (compareFn === undefined) return a.root < z.root ? -1 : 1
-    return compareFn(a, z)
+    if (compareFn !== undefined) return compareFn(a, z)
+
+    if (a.root !== z.root) return a.root < z.root ? -1 : 1
+
+    // SAFETY: variants a and z are of the same `kind` and can only be static or
+    // functional at this point
+
+    let aValue =
+      a.kind === 'static'
+        ? // Static variants don't have a value
+          null
+        : a.kind === 'compound'
+          ? // Compound variants never reach this point
+            null
+          : a.value
+
+    let zValue =
+      z.kind === 'static'
+        ? // Static variants don't have a value
+          null
+        : z.kind === 'compound'
+          ? // Compound variants never reach this point
+            null
+          : z.value
+
+    // Both variants are static, functional without a value, or functional with
+    // equal values
+    if (aValue === zValue) return 0
+
+    // Here both variants must be functional and have different values. One of
+    // the values being `null` here means we have something like `foo` vs
+    // `foo-bar` or `foo-[:bar]` where `foo` is the "root" of the variant.
+    //
+    // We want `foo` to appear before `foo-bar` or `foo-[:bar]`. While this is
+    // an arbitrary decision its one that must be consistent.
+    if (aValue === null) return -1
+    if (zValue === null) return 1
```

Comment:
Yeah this won't ever happen in pratice because:
- In core all our functional variants return null if given no value
- `matchVariant` — which does support a "default" value has a dedicated compare function

Verifying this did reveal that sorting broken there though so I'm fixing that

Reply:
Removed and added a note explaining why this isn't actually necessary.

Reply:
Removing these do result in failure of the postcss tests. I think we should figure out why they are failing there and add them to the `variants.test.ts` if possible.

Maybe as a dedicated sorting test at the bottom of that file?